File: src/main/java/ludwig/model/NodeVisitor.java
package ludwig.model;

public interface NodeVisitor<T> {
    T visitProject(ProjectNode projectNode);

    T visitPackage(PackageNode packageNode);

    T visitFunction(FunctionNode functionNode);

    T visitVariable(VariableNode variableNode);

    T visitAssignment(AssignmentNode assignmentNode);

    T visitReference(ReferenceNode referenceNode);

    T visitList(ListNode listNode);

    T visitLiteral(LiteralNode literalNode);

    T visitLambda(LambdaNode lambdaNode);

    T visitCall(CallNode callNode);

    T visitReturn(ReturnNode returnNode);

    T visitIf(IfNode ifNode);

    T visitElse(ElseNode elseNode);

    T visitFor(ForNode forNode);

    T visitFunctionReference(FunctionReferenceNode functionReference);

    T visitThrow(ThrowNode throwNode);

    T visitPlaceholder(PlaceholderNode placeholderNode);

    T visitBreak(BreakNode breakNode);

    T visitContinue(ContinueNode continueNode);

    T visitOverride(OverrideNode overrideNode);

    T visitClass(ClassNode classNode);
}

File: src/main/java/ludwig/model/NodeVisitor.kt
package ludwig.model

interface NodeVisitor<T> {
    fun visitProject(projectNode: ProjectNode): T

    fun visitPackage(packageNode: PackageNode): T

    fun visitFunction(functionNode: FunctionNode): T

    fun visitVariable(variableNode: VariableNode): T

    fun visitAssignment(assignmentNode: AssignmentNode): T

    fun visitReference(referenceNode: ReferenceNode): T

    fun visitList(listNode: ListNode): T

    fun visitLiteral(literalNode: LiteralNode): T

    fun visitLambda(lambdaNode: LambdaNode): T

    fun visitCall(callNode: CallNode): T

    fun visitReturn(returnNode: ReturnNode): T

    fun visitIf(ifNode: IfNode): T

    fun visitElse(elseNode: ElseNode): T

    fun visitFor(forNode: ForNode): T

    fun visitFunctionReference(functionReference: FunctionReferenceNode): T

    fun visitThrow(throwNode: ThrowNode): T

    fun visitTry(tryNode: TryNode): T

    fun visitPlaceholder(placeholderNode: PlaceholderNode): T

    fun visitBreak(breakNode: BreakNode): T

    fun visitContinue(continueNode: ContinueNode): T

    fun visitOverride(overrideNode: OverrideNode): T

    fun visitClass(classNode: ClassNode): T

    fun visitCatch(catchNode: CatchNode): T
}
